include/mqt-core/qdmi/Driver.hpp

```diff
@@ -0,0 +1,423 @@
+   * timeout.
+   * @see QDMI_job_wait
+   */
+  [[nodiscard]] auto wait(size_t timeout) const -> int;
+
+  /**
+   * @brief Gets the results of the job.
+   * @see QDMI_job_get_results
+   */
+  auto getResults(QDMI_Job_Result result, size_t size, void* data,
+                  size_t* sizeRet) const -> int;
+
+  /**
+   * @brief Frees the job.
+   * @note This function just forwards to the device's @ref Device::freeJob
+   * function. This function is needed because the interface only provides the
+   * job handle to the @ref QDMI_job_free function and the job's device handle
+   * is private.
+   */
+  auto free() -> void;
+};
+
+/**
+ * @brief Definition of the QDMI Session.
+ */
+// Since we treat this struct as a class, we apply also the naming scheme for
+// classes, i.e., an underscore at the end of member names.
+struct QDMI_Session_impl_d {
+private:
+  /// @brief The status of the session.
+  qdmi::SessionStatus status_ = qdmi::SessionStatus::ALLOCATED;
+  /// @brief A pointer to the list of all devices.
+  const std::vector<std::unique_ptr<QDMI_Device_impl_d>>* devices_;
+
+public:
+  /// @brief Constructor for the QDMI session.
+  explicit QDMI_Session_impl_d(
+      const std::vector<std::unique_ptr<QDMI_Device_impl_d>>& devices)
+      : devices_(&devices) {}
+
+  /**
+   * @brief Initializes the session.
+   * @see QDMI_session_init
+   */
+  auto init() -> int;
+
+  /**
+   * @brief Sets a parameter for the session.
+   * @see QDMI_session_set_parameter
+   */
+  auto setParameter(QDMI_Session_Parameter param, size_t size,
+                    const void* value) const -> int;
+
+  /**
+   * @brief Queries a session property.
+   * @see QDMI_session_query_session_property
+   */
+  auto querySessionProperty(QDMI_Session_Property prop, size_t size,
+                            void* value, size_t* sizeRet) const -> int;
+};
+
+namespace qdmi {
+/**
+ * @brief The MQT QDMI driver class.
+ * @details This driver loads all statically known and linked QDMI device
+ * libraries. Additional devices can be added dynamically.
+ * @note This class is a singleton that manages the QDMI libraries and
+ * sessions. It is responsible for loading the libraries, allocating sessions,
+ * and providing access to the devices.
+ */
+class Driver final {
+  /// @brief Private constructor to enforce the singleton pattern.
+  Driver();
+
+  /**
+   * @brief Map of sessions to their corresponding unique pointers to
+   * QDMI_Session_impl_d objects.
+   */
+  std::unordered_map<QDMI_Session, std::unique_ptr<QDMI_Session_impl_d>>
+      sessions_;
+
+  /**
+   * @brief Vector of unique pointers to QDMI_Device_impl_d objects.
+   */
+  std::vector<std::unique_ptr<QDMI_Device_impl_d>> devices_;
+
+public:
+  // Delete copy constructors and assignment operators to prevent copying the
+  // singleton instance.
+  Driver(const Driver&) = delete;
+  Driver& operator=(const Driver&) = delete;
```

Comment:
I am fairly sure this should define (and default) the move and move assignment operators as per the rule of 5.